File: sms-stack-typescript/src/smsTcpController.ts
import { Observable } from 'rxjs';
import { SmsTcpReceiver } from "./smsTcpReceiver";
import { SMSTCPlayer } from "./smsTcpLayer";
import { SmsTcpSender } from "./smsTcpSender";
import { SmsBroadcaster, SmsReceived } from "./index";
import { SmsTcpControllerObserver } from "./smsTcpControllerObserver";
import { SMSTCP } from "./smsTcp";


export class SmsTcpController {
    private receiver: SmsTcpReceiver;
    private sender: SmsTcpSender;
    private cipherMode: number;
    private cipherKey: string;
    private observer: SmsTcpControllerObserver;

    constructor(cipherMode: number = 0, cipherKey: string = "", observer: SmsTcpControllerObserver, broadcasterRec: Observable<SmsReceived>, broadcasterSend: SmsBroadcaster){
        this.cipherMode = cipherMode ;
        this.cipherKey = cipherKey;
        this.receiver = new SmsTcpReceiver(cipherMode, cipherKey, this, broadcasterRec, broadcasterSend);
        this.sender = new SmsTcpSender(cipherMode, cipherKey, this, broadcasterSend); 
        this.observer = observer;
    }


    public handleFinalMessageReceived(messages: string[], sender: string){
        this.observer.handleFinalMessageReceived(messages, sender);
    }

    public handleMessageReceived(layer: SMSTCPlayer, sender: string) {
        this.observer.handleMessageReceived(layer, sender);
    }
    
    public handleFinalMessageSent(messages: string[], sender: string){
        this.observer.handleFinalMessageSent(messages, sender);
    }

    public handleMessageSent(layer: SMSTCPlayer, sender: string) {
        this.observer.handleMessageSent(layer, sender);
    }

    public sendMessage(sms: string, sender: string){
        this.sender.createNewConversation(sms, sender);
    }

    public processMessage(message: string[]): string {
        const finalMessage = message.join("");
        switch(this.cipherMode) {
            case 0:
                return SMSTCP.decodeBase64(finalMessage);
            case 1:
                //TODO - implement AES cipher
                return "";
            default:
                return "";
        }
    }
}

File: sms-stack-python/sms_stack/sms_tcp_controller.py
from .sms_tcp_receiver import SmsTcpReceiver
from .sms_tcp_sender import SmsTcpSender
from .sms_tcp import SmsTcp
from .cipher import AESCipher

class SmsTcpController:

    def __init__(self, observer, broadcaster_send, cipher_mode=0, cipher_key=""):
        """Controller of the framework, runs a sender and receiver to controll the stream of sms
        
        Args:
            observer (SmsTcpObserver): Inherits from SmsControllerObserver implementing the methods to controll the flow of messages
            broadcaster_send (SmsBroadcaster): Class that implements send_message in order to send messages to a sender
            cipher_mode (int, optional): Defaults to 0. Cipher mode of the controller (0 Base64 | 1 AES CTR)
            cipher_key (str, optional): Defaults to "". Pre shared key in the AES cipher
        """
        self.cipher_mode = cipher_mode
        self.cipher_key = cipher_key
        self._observer = observer
        self._receiver = SmsTcpReceiver(cipher_mode, cipher_key, self, broadcaster_send)
        self._sender = SmsTcpSender(cipher_mode, cipher_key, self, broadcaster_send)

    def handle_final_message_received(self, messages, sender):
        self._observer.handle_final_message_received(messages, sender)

    def handle_message_received(self, layer, sender):
        self._observer.handle_message_received(layer, sender)


    def handle_final_message_sent(self, messages, sender):
        self._observer.handle_final_message_sent(messages, sender)

    def handle_message_sent(self, layer, sender):
        self._observer.handle_message_sent(layer, sender)

    def send_message(self, sms, sender):
        self._sender.create_new_conversation(sms, sender)

    def receive_message(self, sms, receiver):
        self._receiver.add_new_message(sms, receiver)


    def process_message(self, message):
        """Process stream of messages received and decode them with the required cipher method
        
        Args:
            message ([str]): List of messages received for the stream
        
        Returns:
            str: Text message decoded and joined
        """

        final_messge = ''.join(message)
        try:
            if(self.cipher_mode == 0):
                return SmsTcp.decode_base_64(final_messge)
            elif(self.cipher_mode == 1):
                aes_cypher = AESCipher()
                clear_text = aes_cypher.decrypt_message(self.cipher_key, final_messge)
                return clear_text
            else:
                return ""
        except Exception as e:
            print(e)
            return ""
